libbpf-tools/lsan.c

```diff
@@ -0,0 +1,1157 @@
+		syms = syms_cache__get_syms(syms_cache, env.pid);
+		if (rst == 0 && syms != NULL)
+			print_report(curr, ip, syms, "direct");
+	}
+
+	/* Report indirect */
+	curr = NULL;
+	next = NULL;
+	HASH_SORT(indirect, report_info_sort);
+	HASH_ITER(hh, indirect, curr, next) {
+		if (count == env.top)
+			break;
+
+		++count;
+		if (curr->id < 0) {
+			printf("%lld bytes indirect leak found in %d allocations from unknown stack\n\n",
+				curr->size * curr->count, curr->count);
+			continue;
+		}
+		rst = bpf_map_lookup_elem(sfd, &(curr->id), ip);
+		syms = syms_cache__get_syms(syms_cache, env.pid);
+		if (rst == 0 && syms != NULL)
+			print_report(curr, ip, syms, "indirect");
+	}
+	free(ip);
+}
+
+static void delete_hash_lsan_hash(struct lsan_info_ext_t *hash) {
+	struct lsan_info_ext_t *curr = NULL;
+	struct lsan_info_ext_t *next = NULL;
+
+	HASH_ITER(hh, hash, curr, next) {
+		HASH_DEL(hash, curr);
+		free(curr);
+	}
+}
+
+static void delete_hash_report_info(struct report_info_t *hash) {
+	struct report_info_t *curr = NULL;
+	struct report_info_t *next = NULL;
+
+	HASH_ITER(hh, hash, curr, next) {
+		HASH_DEL(hash, curr);
+		free(curr);
+	}
+}
+
+static void delete_hash_root_region(struct root_region_t *hash) {
+	struct root_region_t *curr = NULL;
+	struct root_region_t *next = NULL;
+
+	HASH_ITER(hh, hash, curr, next) {
+		HASH_DEL(hash, curr);
+		free(curr);
+	}
+}
+
+static void empty_table()
+{
+	cvector_free(frontier);
+	cvector_free(key_table);
+	cvector_free(tids);
+	delete_hash_lsan_hash(allocs);
+	delete_hash_report_info(direct);
+	delete_hash_report_info(indirect);
+	delete_hash_root_region(certain);
+	delete_hash_root_region(uncertain);
+	frontier = NULL;
+	key_table = NULL;
+	tids = NULL;
+	allocs = NULL;
+	direct = NULL;
+	indirect = NULL;
+	certain = NULL;
+	uncertain = NULL;
+}
+
+static void for_each_tid_ptrace(enum __ptrace_request request)
+{
+	int i = 0;
+	long ret = 0;
+
+	if (!env.stop_the_world)
+		return;
+
+	for (i = 0; i < cvector_size(tids); ++i) {
+		ret = ptrace(request, tids[i], NULL, NULL);
+		if (ret != -1)
+			continue;
+
+		p_warn("ptrace failed to request %d, reason: %s",
+		       request, strerror(errno));
+		p_warn("May failed to stop tid: %d, could cause false alarms",
+		       tids[i]);
+	}
+}
+
+static void for_each_tid_waitpid()
+{
+	int i = 0;
+	int status = 0;
+	pid_t ret = 0;
+
+	if (!env.stop_the_world)
+		return;
+
+	for (i = 0; i < cvector_size(tids); ++i) {
+		ret = waitpid(tids[i], &status, __WALL);
+		if (ret != -1)
+			continue;
+
+		p_warn("waitpid failed, reason: %s", strerror(errno));
+		p_warn("May failed to stop tid: %d, could cause false alarms",
+		       tids[i]);
+	}
+}
+
+static int do_leak_check(struct syms_cache *syms_cache)
+{
+	int ret = 0;
+	int len = 0;
+
+	cvector_reserve(tids, 1024);
+	len = get_tids(env.pid, cvector_begin(tids), 1024);
+	cvector_set_size(tids, len);
+	for_each_tid_ptrace(PTRACE_SEIZE);
+	for_each_tid_ptrace(PTRACE_INTERRUPT);
+	for_each_tid_waitpid();
```

Comment:
Yes, it could be possible to have the code inside one loop and doing all `ptrace` `attach`, `interrupt` and `waitpid` inside the loop, which will mitigate loop overhead.

I think it would be better to factorize it for readability.
However, despite the intention to factorize it for readability, it currently doesn't seem to be very readable.
I'll revise this function to enhance its readability.

Thanks!

Reply:
I tried to change the code as you suggested, but because of the `env.stop_the_world` flag, the code became more complicated. This means the current code has better readability in my opinion. So I decided to stick with it. However, your suggestion is very good, and I'll change the code if I find a more readable way.

Thanks!